Record how many parameters we're parsing within dynamo

c00w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148508

This allows us to track how many paramaters we have in compilations. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

c00w committed:
WIP: record how many paramaters we're parsing

[ghstack-poisoned]

c00w committed:
Update on "WIP: record how many paramaters we're parsing"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "WIP: record how many paramaters we're parsing"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "WIP: record how many paramaters we're parsing"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "WIP: record how many paramaters we're parsing"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "WIP: record how many paramaters we're parsing"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. Similar to https://github.com/pytorch/pytorch/pull/147149 this is a total counter, not a per compile counter.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
None on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

XuehaiPan commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1511,6 +1511,14 @@ def mark_static_input(self, value: torch.Tensor, guard: bool):
         from ..eval_frame import OptimizedModule
 
+        try:
+            params = list(value.parameters(recurse=False))
```
Maybe should also handle tied weights?

c00w replied:
Good question - I agree this will overcount with tied weights. Do you have any thoughts on how to fix that?

I guess we could store like a cache of "already counted" tensor paramaters, and if we've already counted it once, not include it again.

XuehaiPan replied:
> I guess we could store like a cache of "already counted" tensor paramaters, and if we've already counted it once, not include it again.

I think this can be achieved by having a weakref set.

c00w replied:
Done.

jansel commented on torch/_dynamo/utils.py:
```diff
@@ -1230,6 +1230,7 @@ class CompilationMetrics:
     recompile_reason: Optional[str] = None
     num_graph_breaks: Optional[int] = None
     triton_kernel_compile_times_us: Optional[str] = None
+    num_params: Optional[int] = None
```
I'd expect this to be a count of tensors, not of numel.  Do we want numel or bytes?

XuehaiPan replied:
> I'd expect this to be a count of tensors, not of numel. Do we want numel or bytes?

It is a common practice to make the number of parameters refer to:

```python
sum(p.numel() for _, p in model.named_parameters(recurse=True, remove_duplicate=True))
```

Maybe we can also add a new metric for number of tensors (e.g., `num_vars`).

c00w replied:
@jansel - We actually want the size of the paramaters. We're trying to measure how large the model is. Zero objection to also recording how many individual parameters there are, but here I explicitly wanted to use numel :). (and my understanding agrees with @XuehaiPan  - which is normally # of paramaters is how many individual elements there are within the tensors, not how large they are - i.e. a fp8 and fp32 model with the same architecture and dimensions, would have the same number of paramaters).

c00w replied:
Nice catch re: remove duplicates. Let me fix that.

jansel replied:
Let's include numel in the name in case we need to add bytes in the future.

c00w replied:
Done.

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented:
Added all 3 new params, need to debug the test failures.

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

XuehaiPan commented on test/dynamo/test_utils.py:
```diff
@@ -484,6 +490,52 @@ def test2(x):
+            "log_compilation_metrics": True,
+        }
+    )
+    def test_num_params(self):
```
Please add a test for tied modules and tied weights.

```python
@torch.compile
def func(m1, m2, x):
    return m1(x) + m2(x)

x = torch.randn(...)
m = nn.Linear(...)
m1 = nn.Linear(...)
m2 = nn.Linear(...)

# Pass same module
func(m, m, x)

# Pass different module
func(m1, m2, x)

m2.weight = m1.weight
# Tie weight
func(m1, m2, x)
```

------

I'm unsure of the correct expected result of when there are tied weights split into different graphs. cc @jansel for thoughts?

```python
def func(m1, m2_with_tie_weight, x):
    y = m1(x)
    torch._dynamo.graph_break()
    z = m2_with_tie_weight(y)
    return z
```

c00w replied:
re: split into seperate graphs, I expect we'll want to double count them, since these metrics are being exported per compile event (aka per graph).

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jansel reviewed (changes requested):
Looks like this is a compile time regression, we should find out how to do this without hurting compile time.

XuehaiPan commented on test/dynamo/test_utils.py:
```diff
@@ -484,6 +490,76 @@ def test2(x):
+        with mock.patch("torch._dynamo.utils.log_compilation_event") as log_event:
+            func(m1, m2, torch.randn(4, 4))
+            compilation_events = [arg[0][0] for arg in log_event.call_args_list]
+        self.assertEqual(compilation_events[0].param_numel, 40)
+        self.assertEqual(compilation_events[0].param_bytes, 160)
+        self.assertEqual(compilation_events[0].param_count, 4)
```
The tied weight test is not correct.

```suggestion
        self.assertEqual(compilation_events[0].param_numel, 24)
        self.assertEqual(compilation_events[0].param_bytes, 96)
        self.assertEqual(compilation_events[0].param_count, 3)
```

XuehaiPan commented on test/dynamo/test_utils.py:
```diff
@@ -484,6 +490,76 @@ def test2(x):
+        self.assertEqual(compilation_events[0].param_bytes, 4 * 1040)
+        self.assertEqual(compilation_events[0].param_count, 4)
+
+        @torch.compile()
+        def func(m1, m2, x):
+            return m1(x) + m2(x)
+
+        # Test a tied module
+        m = nn.Linear(4, 4)
+        with mock.patch("torch._dynamo.utils.log_compilation_event") as log_event:
+            func(m, m, torch.randn(4, 4))
+            compilation_events = [arg[0][0] for arg in log_event.call_args_list]
+        self.assertEqual(compilation_events[0].param_numel, 20)
+        self.assertEqual(compilation_events[0].param_bytes, 80)
+        self.assertEqual(compilation_events[0].param_count, 2)
+
+        # Test tied weights
+        m1 = nn.Linear(4, 4)
+        m2 = nn.Linear(4, 4)
+        m1.weight = m2.weight
+        with mock.patch("torch._dynamo.utils.log_compilation_event") as log_event:
+            func(m1, m2, torch.randn(4, 4))
+            compilation_events = [arg[0][0] for arg in log_event.call_args_list]
+        self.assertEqual(compilation_events[0].param_numel, 40)
+        self.assertEqual(compilation_events[0].param_bytes, 160)
+        self.assertEqual(compilation_events[0].param_count, 4)
```
We can create nested test module with:

```suggestion
        # Test a tied module
        l1 = nn.Linear(4, 4)
        l2 = nn.Linear(4, 4)
        m = nn.Sequential(l1, nn.Sequential(l1, l2))
        self.assertEqual([x.numel() for x in m.parameters()], [16, 4, 16, 4])
        with mock.patch("torch._dynamo.utils.log_compilation_event") as log_event:
            torch.compile(m)(torch.randn(4, 4))
            compilation_events = [arg[0][0] for arg in log_event.call_args_list]
        self.assertEqual(compilation_events[0].param_numel, 40)
        self.assertEqual(compilation_events[0].param_bytes, 160)
        self.assertEqual(compilation_events[0].param_count, 4)

        # Test tied weights
        l1 = nn.Linear(4, 4)
        l2 = nn.Linear(4, 4)
        l2.weight = l1.weight
        m = nn.Sequential(l1, nn.Sequential(l2))
        self.assertEqual([x.numel() for x in m.parameters()], [16, 4, 4])
        with mock.patch("torch._dynamo.utils.log_compilation_event") as log_event:
            torch.compile(m)(torch.randn(4, 4))
            compilation_events = [arg[0][0] for arg in log_event.call_args_list]
        self.assertEqual(compilation_events[0].param_numel, 24)
        self.assertEqual(compilation_events[0].param_bytes, 96)
        self.assertEqual(compilation_events[0].param_count, 3)
```

c00w replied:
Updated tied module :).

c00w commented:
@jansel  - Where do you see the compile time regression?

jansel commented:
It was failing (>1.5%) last CI run, it looks like now it is at 1.4% but still a regression:
https://github.com/pytorch/pytorch/actions/runs/13818825985/job/38669115043?pr=148508#step:22:836

c00w committed:
None on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
None on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented:
re: Tied weights etc...
Changed the metrics context to store this per paramater (and using a ID instead of a weakref), this should resolve issues with tied wieghts. @XuehaiPan  - Let me know if you spot any more bugs that you think are fixable.

re: compile time count regression - AFAICT, there is now way to 'mitigate" this since we're just counting how many instructions we run, and iterating through a paramater ads a lot of instructions. @jansel  - let me know what your thought is here. I've modified it to only run when we're within a metrics context, but if you want I can be more strict and only run it when log_compilation_metrics is enabled (or maybe even only log it inside fbcode).

Please let me know if I'm misunderstanding the compile time instruction count benchmark here.

XuehaiPan commented on test/dynamo/test_utils.py:
```diff
@@ -489,6 +495,79 @@ def test2(x):
+        self.assertEqual(compilation_events[0].param_bytes, 4 * 1040)
+        self.assertEqual(compilation_events[0].param_count, 4)
+
+        @torch.compile()
+        def func(m1, m2, x):
+            return m1(x) + m2(x)
```
We can remove this. See my suggestion in https://github.com/pytorch/pytorch/pull/148508#discussion_r1992228443.

c00w replied:
Ah sorry - Fixed (hand coppied it and missed the change to directly compile m.

XuehaiPan commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1516,6 +1516,27 @@ def mark_static_input(self, value: torch.Tensor, guard: bool):
+                    )
+                    if not torch.nn.parameter.is_lazy(p)
+                )
+                metrics_context = get_metrics_context()
+                for p in params:
+                    metrics_context.set_key_value("param_numel_addr", id(p), p.numel())
+                    metrics_context.set_key_value("param_bytes_addr", id(p), p.nbytes)
+                    metrics_context.set_key_value("param_count_addr", id(p), 1)
```
Change generator expression `params = (...)` to list comprehension `params = [...]`. Then move this out of the `try`-clause.

c00w replied:
Done.

XuehaiPan commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1516,6 +1516,27 @@ def mark_static_input(self, value: torch.Tensor, guard: bool):
+                )
+                metrics_context = get_metrics_context()
+                for p in params:
+                    metrics_context.set_key_value("param_numel_addr", id(p), p.numel())
```
I suspect adding a new member to `class MetricsContext`.

```python
class MetricsContext:
    def __init__(self, on_exit: OnExitType):
        self._on_exit = on_exit
        self._metrics: dict[str, Any] = {}
        self._traced_parameters: WeakSet[nn.Parameter] = WeakSet()  # <==
        self._start_time_ns: int = 0
        self._level: int = 0

    def track(self, param: nn.Parameter) -> bool:
        tracked = param in self._traced_parameters
        self._traced_parameters.add(param)
```

c00w replied:
Switched.

c00w replied:
FYI - appears paramaters don't work with WeakSets, still had to use id(param).
```
025-03-28T02:01:09.9298631Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_dynamo/metrics_context.py", line 147, in track
2025-03-28T02:01:09.9299254Z     tracked = param in self._traced_parameters
2025-03-28T02:01:09.9299775Z   File "/opt/conda/envs/py_3.10/lib/python3.10/_weakrefset.py", line 80, in __contains__
2025-03-28T02:01:09.9300268Z     return wr in self.data
2025-03-28T02:01:09.9300893Z torch._dynamo.exc.InternalTorchDynamoError: RuntimeError: Boolean value of Tensor with more than one value is ambiguous
2025-03-28T02:01:09.9301438Z 
```

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented on torch/_dynamo/config.py:
```diff
@@ -600,6 +600,9 @@ def default_debug_dir_root():
     env_name_default="TORCH_DYNAMO_RUN_GC_AFTER_COMPILE",
 )
 
+# Whether to profile dynamo modules when we run.
```
Sorry removed, this was a half finished thought - Instead just made this depend on the log_compilation_event config.

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1516,6 +1516,27 @@ def mark_static_input(self, value: torch.Tensor, guard: bool):
 
+        # if config.profile_params:
+        metrics_context = get_metrics_context()
+        if metrics_context.in_progress():
```
Removed.

c00w commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1516,6 +1516,27 @@ def mark_static_input(self, value: torch.Tensor, guard: bool):
+                )
+                metrics_context = get_metrics_context()
+                for p in params:
+                    metrics_context.set_key_value("param_numel_addr", id(p), p.numel())
```
We're using this as a de duplication method, added a comment to make this clear.

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1516,6 +1516,27 @@ def mark_static_input(self, value: torch.Tensor, guard: bool):
     def wrap_module(self, value: torch.nn.Module):
         from ..eval_frame import OptimizedModule
 
+        # if config.profile_params:
```
Sorry in progress push - removed.

c00w commented on torch/_dynamo/utils.py:
```diff
@@ -1317,6 +1324,19 @@ def collection_to_json_str(metric: Optional[Any]) -> Optional[str]:
         compile_id = all_metrics.get("compile_id")
         all_metrics["compile_id"] = str(compile_id) if compile_id else None
 
+        all_metrics["param_numel"] = sum(
```
Added a comment.

c00w commented on torch/_dynamo/utils.py:
```diff
@@ -1234,6 +1234,13 @@ class CompilationMetrics:
     ir_count: Optional[int] = None
     cudagraph_skip_reason: Optional[str] = None
     python_version: Optional[str] = None
+    # The number of elements within paramaters. This is classically what people think of when they think of paramaters
```
Fixed. (Or at least validated with sl show . | grep paramater)

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented:
FYI - Tests may fail (still waiting for PR to build), but everything else should be resolved, and wanted to get feedback faster.

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Record how many paramaters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral commented on test/dynamo/test_utils.py:
```diff
@@ -489,6 +495,76 @@ def test2(x):
+        with mock.patch("torch._dynamo.utils.log_compilation_event") as log_event:
+            torch.compile(m)(torch.randn(4, 4))
+            compilation_events = [arg[0][0] for arg in log_event.call_args_list]
+        self.assertEqual(compilation_events[0].param_numel, 24)
```
@XuehaiPan I'm trying to help resurrect this PR while c00w is out. There are several things that look wrong here, wondering if you can help since you were more involved in this PR.

1) The logic to track the id's (the "clever" part below) doesn't work as expected. I'm getting non-deterministic results here because sometimes Python is reusing an id from a previous compilation; presumably because the module is GC'd since we're reusing 'm'.

2) I'm not sure the "clever" part is needed anyway. If I test each of these sections in isolation, I'm never seeing an already-tracked id for either of the "tied module" or "tied weights" cases. It would seem that querying the params with `remove_duplicate=True` is already providing the functionality we want here?

3) What values would you actually expect here? I think 24, 96, 3 are wrong.

masnesral committed:
Update on "Record how many parameters we're parsing within dynamo"


This allows us to track how many paramaters we have in compilations. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

masnesral commented on torch/_dynamo/guards.py:
```diff
@@ -2141,6 +2141,12 @@ def TENSOR_MATCH(self, guard: Guard, value=None):
             value = value if value is not None else self.get(guard.name)
             assert isinstance(value, torch.Tensor)
 
+            if config.log_compilation_metrics and isinstance(value, torch.nn.Parameter):
```
Changed the implementation per offline discussion with @anijain2305. Animesh, do you mind having a look?

masnesral commented:
@pytorchbot merge
